File: 21.合并两个有序链表.cpp
/*
 * @lc app=leetcode.cn id=21 lang=cpp
 *
 * [21] 合并两个有序链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {

        ListNode * head = new ListNode(0);
        ListNode * p = head;

        while (l1 != NULL && l2 != NULL){

            if (l1->val <= l2->val) p->next = l1, l1 = l1->next;
            else  {
                p->next = l2;
                l2 = l2->next;
            }
            p = p->next;

        }

        // 把剩下的链给接上
        if (l1 != NULL) p->next = l1;
        if (l2 != NULL) p->next = l2;
        
        return head->next; // 返回不带头节点的 链表;
    }
};
// @lc code=end

File: 21.合并两个有序链表.java
import java.util.PriorityQueue;

/*
 * @lc app=leetcode.cn id=21 lang=java
 *
 * [21] 合并两个有序链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {

       if (l1 == null) return l2;
       if (l2 == null) return l1;

       ListNode ans = new ListNode(0);
       ListNode temp = ans;

       while (l1 != null && l2 != null) {
            if (l1.val > l2.val) {
                temp.next = l2;
                l2 = l2.next;
            } else {
                temp.next = l1;
                l1 = l1.next;
            }
            temp = temp.next;
       }
       if (l1 != null) temp.next = l1;
       if (l2 != null) temp.next = l2;

        return ans.next;

    }
}
// @lc code=end

